syntax = "proto3";

service KeyValueStore {
    rpc Get (GetKey) returns (GetReply) {}
    rpc Put (PutKey) returns (PutReply) {}
    rpc List (Void) returns (StoreReply) {}
    rpc RegisterWithPeer (IP) returns (StoreReply) {}
}

message GetKey {
    string key = 1;
}

message GetReply {
    string value = 1;
    bool defined = 2;
}

message PutKey {
    string key = 1;
    string value = 2;
    bool broadcast = 3;
}

message PutReply {
    string value = 1;
}

message Void {
}

message IP {
    string ip = 1;
}

message StoreReply {
    map<string, string> store = 1;
}

option py_generic_services = true;